src/app_client.rs

```diff
@@ -59,45 +65,76 @@ fn data_cells_from_rows(rows: Vec<Option<Vec<u8>>>) -> Result<Vec<DataCell>> {
 
 #[instrument(skip_all, fields(block = block.block_num), level = "trace")]
 async fn process_block(
+	cfg: &AppClientConfig,
 	db: Arc<DB>,
+	network_client: Client,
 	rpc_url: &str,
 	app_id: u32,
 	block: &ClientMsg,
 	pp: PublicParameters,
 ) -> Result<()> {
+	let lookup = &block.lookup;
 	let block_number = block.block_num;
+	let dimensions = &block.dimensions;
 	let commitments = &block.commitment;
 
-	let rows = get_kate_app_data(rpc_url, block.header_hash, app_id).await?;
-	let rows_count = rows.iter().filter(|&o| Option::is_some(o)).count();
+	let mut rows = get_kate_app_data(rpc_url, block.header_hash, app_id).await?;
+
+	let rows_count = rows.iter().filter(|row| row.is_some()).count();
 	info!(block_number, "Found {rows_count} rows for app {app_id}");
 
-	let is_verified = kate_recovery::commitments::verify_equality(
-		&pp,
-		commitments,
-		&rows,
-		&block.lookup,
-		&block.dimensions,
-		app_id,
-	)?;
+	let (verified_rows, missing_rows) =
+		commitments::verify_equality(&pp, &commitments, &rows, &lookup, &dimensions, app_id)?;
+
+	info!(block_number, "Verified rows: {verified_rows:?}");
 
-	info!(block_number, "Block verified: {is_verified}");
+	let mut have_missing = !missing_rows.is_empty();
 
-	if is_verified {
+	if have_missing {
+		let missing_positions = dimensions.extended_rows_positions(&missing_rows);
+		let (fetched, unfetched) = fetch_cells_from_dht(
+			&network_client,
+			block_number,
+			&missing_positions,
+			cfg.dht_parallelization_limit,
+		)
+		.await
+		.context("Failed to fetch cells from DHT")?;
```

Comment:
We should a time spent on dht cell retrieval metric here, at some point...

Reply:
Yeah, I hope ill not forget to add metrics in next iteration (column reconstruction)